src/dra/7E4BC.c

```diff
@@ -1235,7 +1235,435 @@ Primitive* func_80121F58(bool arg0, s32 arg1, Primitive* arg2, s16 facingLeft) {
+    for (j = 0; j < 16; j++) {
+        xVar5 = D_80138094[j].posX.i.hi;
+        yVar5 = D_80138094[j].posY.i.hi;
+        xVar2 = D_80138094[(j + 1) % 16].posX.i.hi;
+        yVar2 = D_80138094[(j + 1) % 16].posY.i.hi;
+        prim->x0 = xVar5;
+        prim->y0 = yVar5;
+        prim->x1 = xVar2;
+        prim->y1 = yVar2;
+        prim->x2 = xVar4;
+        prim->x3 = xVar4;
+        prim->y2 = yVar4;
+        prim->y3 = yVar4;
+        prim->r0 = prim->g0 = prim->b0 = prim->r1 = prim->g1 = prim->b1 = 0;
+        prim->r2 = prim->g2 = prim->b2 = prim->r3 = prim->g3 = prim->b3 =
+            D_80138398;
+#ifndef VERSION_HD
+        if (D_80138398 < 0x10) {
+            prim->drawMode |= DRAW_HIDE;
+        } else {
+            prim->drawMode &= ~DRAW_HIDE;
+        }
+#endif
+        prim = prim->next;
+    }
+    for (j = 0; j < 16; j++) {
+        xVar5 = D_80138094[j].posX.i.hi;
+        yVar5 = D_80138094[j].posY.i.hi;
+        xVar2 = D_80138094[(j + 1) % 16].posX.i.hi;
+        yVar2 = D_80138094[(j + 1) % 16].posY.i.hi;
+        for (i = 0; i < 4; i++) {
+            prim->x0 = xVar4 + (i + 1) * (xVar5 - xVar4) / 4;
+            prim->y0 = yVar4 + (i + 1) * (yVar5 - yVar4) / 4;
+            prim->x1 = xVar4 + (i + 1) * (xVar2 - xVar4) / 4;
+            prim->y1 = yVar4 + (i + 1) * (yVar2 - yVar4) / 4;
+            prim->x2 = xVar4 + i * (xVar5 - xVar4) / 4;
+            prim->y2 = yVar4 + i * (yVar5 - yVar4) / 4;
+            prim->x3 = xVar4 + i * (xVar2 - xVar4) / 4;
+            prim->y3 = yVar4 + i * (yVar2 - yVar4) / 4;
+            switch (i) {
+            case 0:
+                prim->r0 = prim->g0 = prim->b0 = prim->r1 = prim->g1 =
+                    prim->b1 = (D_80138394 * primColor2 / 256) & 0xFF;
+                prim->r2 = prim->g2 = prim->b2 = prim->r3 = prim->g3 =
+                    prim->b3 = (D_80138394 * primColor1 / 256) & 0xFF;
+                break;
+            case 1:
+                prim->r0 = prim->g0 = prim->b0 = prim->r1 = prim->g1 =
+                    prim->b1 = (D_80138394 * primColor3 / 256) & 0xFF;
+                prim->r2 = prim->g2 = prim->b2 = prim->r3 = prim->g3 =
+                    prim->b3 = (D_80138394 * primColor2 / 256) & 0xFF;
+                break;
+            case 2:
+                prim->r0 = prim->g0 = prim->b0 = prim->r1 = prim->g1 =
+                    prim->b1 = (D_80138394 * primColor4 / 256) & 0xFF;
+                prim->r2 = prim->g2 = prim->b2 = prim->r3 = prim->g3 =
+                    prim->b3 = (D_80138394 * primColor3 / 256) & 0xFF;
+                break;
+            case 3:
+                prim->r0 = prim->g0 = prim->b0 = prim->r1 = prim->g1 =
+                    prim->b1 = 0;
+                prim->r2 = prim->g2 = prim->b2 = prim->r3 = prim->g3 =
+                    prim->b3 = (D_80138394 * primColor4 / 256) & 0xFF;
+                break;
+            }
+            prim = prim->next;
+        }
+    }
+
+block_147:
```

Comment:
Tricky business... Only the first two `goto` can be replaced with:
```c
self->facingLeft = PLAYER.facingLeft;
return;
```
I am not sure if you want to remove those two `goto` or be consistent and keep all of them falling to the last function statement

Reply:
I think I will keep it this way at least for now. PSP seems to want to use the goto, so I think unfortunately the goto might be real.